File: mobile/android/app/src/main/kotlin/dev/activechain/wallet/AgentWallet.kt
package dev.activechain.wallet

enum class AgentConnection(val label: String) {
    WALLET_OWNED("Wallet-owned"),
    THIRD_PARTY("Third-party"),
    REMOTE("Remote"),
    MANAGED_DEVICE("Managed device")
}

sealed class AgentLifecycle {
    data object Active : AgentLifecycle()
    data object Paused : AgentLifecycle()
    data object RevocationPending : AgentLifecycle()
    data class Revoked(val finalizedHeight: Long) : AgentLifecycle()
}

data class AgentDelegation(val id: String, val label: String, val capabilities: List<String>,
                           val dailyLimit: Long, val expiresAt: Long,
                           val connection: AgentConnection = AgentConnection.THIRD_PARTY,
                           var spentToday: Long = 0,
                           var lifecycle: AgentLifecycle = AgentLifecycle.Active) {
    val revoked: Boolean get() = lifecycle is AgentLifecycle.Revoked
}
data class PendingApproval(val id: String, val agentId: String, val recipient: String,
                           val amount: Long, val feeReserve: Long, val networkId: String)

class AgentWalletStore {
    val agents = mutableListOf<AgentDelegation>()
    val pending = mutableListOf<PendingApproval>()
    fun delegate(agent: AgentDelegation): Boolean { if (agents.any { it.id == agent.id }) return false; agents += agent; return true }
    fun enqueue(approval: PendingApproval) { pending += approval }
    fun pause(id: String) {
        agents.firstOrNull { it.id == id && it.lifecycle == AgentLifecycle.Active }?.lifecycle =
            AgentLifecycle.Paused
    }
    fun resume(id: String) {
        agents.firstOrNull { it.id == id && it.lifecycle == AgentLifecycle.Paused }?.lifecycle =
            AgentLifecycle.Active
    }
    fun revoke(id: String) {
        agents.firstOrNull { it.id == id && it.lifecycle !is AgentLifecycle.Revoked }?.lifecycle =
            AgentLifecycle.RevocationPending
    }
    fun finalizeRevocation(id: String, height: Long) {
        if (height <= 0) return
        agents.firstOrNull {
            it.id == id && it.lifecycle == AgentLifecycle.RevocationPending
        }?.lifecycle = AgentLifecycle.Revoked(height)
    }
}

File: mobile/ios/ActiveChainWalletApp/Sources/AgentWallet.swift
import Foundation
import LocalAuthentication

public enum AgentConnection: String, Equatable {
    case walletOwned = "Wallet-owned"
    case thirdParty = "Third-party"
    case remote = "Remote"
    case managedDevice = "Managed device"
}

public enum AgentLifecycle: Equatable {
    case active
    case paused
    case revocationPending
    case revoked(finalizedHeight: UInt64)
}

public struct AgentDelegation: Identifiable, Equatable {
    public let id: String
    public let label: String
    public let capabilities: [String]
    public let dailyLimit: UInt64
    public let expiresAt: UInt64
    public let connection: AgentConnection
    public var spentToday: UInt64
    public var lifecycle: AgentLifecycle

    public init(id: String, label: String, capabilities: [String], dailyLimit: UInt64,
                expiresAt: UInt64, revoked: Bool = false,
                connection: AgentConnection = .thirdParty, spentToday: UInt64 = 0,
                lifecycle: AgentLifecycle? = nil) {
        self.id = id
        self.label = label
        self.capabilities = capabilities
        self.dailyLimit = dailyLimit
        self.expiresAt = expiresAt
        self.connection = connection
        self.spentToday = spentToday
        self.lifecycle = lifecycle ?? (revoked ? .revoked(finalizedHeight: 1) : .active)
    }

    public var revoked: Bool {
        if case .revoked = lifecycle { return true }
        return false
    }
}

public final class BiometricAuthorizer {
    public init() {}

    /// Requests biometrics only; callers must fail closed on any error and never invoke
    /// `deviceOwnerAuthentication` as a passcode fallback for high-value actions.
    public func authorize(reason: String, completion: @escaping (Bool, Error?) -> Void) {
        let context = LAContext()
        context.localizedFallbackTitle = ""
        context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason) { success, error in
            DispatchQueue.main.async { completion(success, error) }
        }
    }
}
